kyo-data/shared/src/main/scala/kyo/Record.scala

```diff
@@ -0,0 +1,642 @@
+        def asRecord(value: A): Record[Fields]
+    end AsRecord
+
+    object AsRecord:
+
+        type FieldsOf[Names <: Tuple, Values <: Tuple] = Names match
+            case nHead *: EmptyTuple => Values match
+                    case vHead *: _ => (nHead ~ vHead)
+            case nHead *: nTail => Values match
+                    case vHead *: vTail => (nHead ~ vHead) & FieldsOf[nTail, vTail]
+            case _ => Any
+
+        type RMirror[A, Names <: Tuple, Values <: Tuple] = Mirror.ProductOf[A] {
+            type MirroredElemLabels = Names
+            type MirroredElemTypes  = Values
+        }
+
+        trait RecordContents[Names <: Tuple, Values <: Tuple]:
+            def values(product: Tuple): List[(Field[?, ?], Any)]
+
+        object RecordContents:
+            given empty: RecordContents[EmptyTuple, EmptyTuple] with
+                def values(product: Tuple): List[(Field[?, ?], Any)] = Nil
+
+            given nonEmpty[NH <: (String & Singleton), NT <: Tuple, VH, VT <: Tuple](
+                using
+                tag: Tag[VH],
+                vo: ValueOf[NH],
+                next: RecordContents[NT, VT]
+            ): RecordContents[NH *: NT, VH *: VT] with
+                def values(product: Tuple): List[(Field[?, ?], Any)] =
+                    (Field[NH, VH](vo.value, tag), product.head) +: next.values(product.tail)
+            end nonEmpty
+        end RecordContents
+
+        given [A <: Product, Names <: Tuple, Values <: Tuple](using
+            mir: RMirror[A, Names, Values],
+            rc: RecordContents[Names, Values]
+        ): AsRecord[A] with
+            type Fields = FieldsOf[Names, Values]
+
+            def asRecord(value: A): Record[Fields] =
+                val record_contents = rc.values(Tuple.fromProduct(value))
+                val map             = Map(record_contents*)
+                map.asInstanceOf[Record[Fields]]
+            end asRecord
+        end given
+    end AsRecord
+
+    /** Represents a field in a Record type at the type level.
+      *
+      * AsField is used to convert Record field types into concrete Field instances, maintaining type safety for field names and their
+      * associated values.
+      *
+      * @tparam A
+      *   The field type, typically in the form of `"fieldName" ~ ValueType`
+      */
+    opaque type AsField[+A] = Field[?, ?]
+
+    object AsField:
+        inline given [N <: String, V](using tag: Tag[V]): AsField[N ~ V] =
+            Field(constValue[N], tag)
+
+        def apply[A](using af: AsField[A]): Field[?, ?] = af
+    end AsField
+
+    /** Type class for working with sets of Record fields.
+      *
+      * AsFields provides type-safe field set operations and is used primarily for the `compact` operation on Records.
+      *
+      * @tparam A
+      *   The combined type of all fields in the set
+      */
+    opaque type AsFields[+A] <: Set[Field[?, ?]] = Set[Field[?, ?]]
+
+    trait AsFields22:
```

Comment:
I tried a few approaches but I couldn't make the compiler destructure the type intersections so I decided to go with brute force